tracing/src/main/java/com/palantir/remoting3/tracing/Tracer.java

```diff
@@ -123,23 +126,40 @@ private static OpenSpan startSpanInternal(String operation, SpanType type) {
     }
 
     /**
+     * Completes the current span (if it exists) and notifies all {@link #observers subscribers} about the
+     * completed span.
+     *
+     * Does not construct the Span object if no subscriber will see it.
+     */
+    public static void fastCompleteSpan() {
+        completeSpanInternal(Collections.emptyMap(), false);
+    }
+
+    /**
+     * Like {@link #fastCompleteSpan()}, but adds {@code metadata} to the current span being completed.
+     */
+    public static void fastCompleteSpan(Map<String, String> metadata) {
+        completeSpanInternal(metadata, false);
+    }
+
+    /**
      * Completes and returns the current span (if it exists) and notifies all {@link #observers subscribers} about the
      * completed span.
      */
     public static Optional<Span> completeSpan() {
-        return completeSpanInternal(Collections.emptyMap());
+        return completeSpanInternal(Collections.emptyMap(), true);
     }
 
     /**
      * Like {@link #completeSpan()}, but adds {@code metadata} to the current span being completed.
      */
     public static Optional<Span> completeSpan(Map<String, String> metadata) {
-        return completeSpanInternal(metadata);
+        return completeSpanInternal(metadata, true);
     }
 
-    private static Optional<Span> completeSpanInternal(Map<String, String> metadata) {
+    private static Optional<Span> completeSpanInternal(Map<String, String> metadata, boolean mustConstructResult) {
         Optional<OpenSpan> maybeOpenSpan = currentTrace.get().pop();
```

Comment:
so let's make:

    private static Optional<OpenSpan> popCurrentSpan() {
        return currentTrace.get().pop();
    }

Then:

    private static void fastCompleteSpan() {
        popCurrentSpan();
    }

and `completeSpan` and `completeSpanInternal` should largely merge back together.

Reply:
this also lets you make `fastCompleteSpan(Map<String, String> garbage)` something that can ignore `garbage`.

Reply:
also, why does `fastCompleteSpan(Map<String, String>)` exist as a thing in this construct?

Reply:
fastCompleteSpan doesn't just want to do this though - it also needs to talk to subscribers

Reply:
however your meta-point makes sense and I'll refactor the code into something that makes more sense